capability defaulting

pweil- opened the pull request:
@smarterclayton @eparis - here is the WIP for adding required add/drops into SCCs.  PTAL

The mechanics are:

1.  You may define `requiredCapabilities` in the SCC which can include adds and drops
1.  Defaulting will produce
     1.  Adds: all required caps + container requested adds - container requested drops (you can drop required adds if you want, just means your container can do less)
     1.  Drops: all required drops + container requested drops
1.  Validation will ensure
     1.  Adds: all adds must be in either the required or allowed add lists of the SCC
     1.  Drops:  all required drops are in the drop set

smarterclayton commented:
Do we need to be smart and not add defaults when they are already dropped (by the container or scc)?

pweil- commented:
> Do we need to be smart and not add defaults when they are already dropped (by the container or scc)?

Since I'm using a string set and using the union it will not duplicate the adds.  If the container is dropping something that is added by default it will be removed by the `Difference` call here: https://github.com/openshift/origin/pull/6470/files#diff-c7bf63e9488690fe18014ead17dbdd2cR61

smarterclayton commented:
[test]

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2597,7 +2597,14 @@ type SecurityContextConstraints struct {
+	// the containers.  Any capability listed in add will be added to the container even if
+	// it is not requested.  Any capability listed in drop will be removed from the container
+	// even if it is requested to be added.
+	RequiredCapabilities Capabilities `json:"requiredCapabilities" description:"capabilities that must be added and dropped"`
```
`added or dropped`

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2597,7 +2597,14 @@ type SecurityContextConstraints struct {
+	// Required capabilities are the capability set that must be set or defaulted into
+	// the containers.  Any capability listed in add will be added to the container even if
+	// it is not requested.  Any capability listed in drop will be removed from the container
+	// even if it is requested to be added.
```
If someone requests a capability to be added explicitly, and it is in the drop list for required capabilities, I would expect the pod to get rejected with a failure, not accepted with the capability dropped.

jwhonce commented:
@twiest FYI

smarterclayton commented:
One more question.  An admin wants to run a pod.  They have access to various high privilige sccs.  When they run a pod that wants to ping, will they get defaulted to drop raw net caps?

pweil- commented:
> One more question. An admin wants to run a pod. They have access to various high privilige sccs. When they run a pod that wants to ping, will they get defaulted to drop raw net caps?

In this case, the assumption would be that the admin has access to an SCC that does not require the net caps to be dropped.  The admin should use prioritization to ensure that they get the defaults they expect when not requesting specific capabilities that would force validation failures (similar to requesting UID 0 before we could prioritize the `anyuid` SCC).

smarterclayton commented:
Ok - those changes will be follow up / split PR?

On Mon, Jan 11, 2016 at 2:52 PM, Paul Weil <notifications@github.com> wrote:

> One more question. An admin wants to run a pod. They have access to
> various high privilige sccs. When they run a pod that wants to ping, will
> they get defaulted to drop raw net caps?
>
> In this case, the assumption would be that the admin has access to an SCC
> that does not require the net caps to be dropped. The admin should use
> prioritization to ensure that they get the defaults they expect when not
> requesting specific capabilities that would force validation failures
> (similar to requesting UID 0 before we could prioritize the anyuid SCC).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6470#issuecomment-170669711>.
>

pweil- commented:
> Ok - those changes will be follow up / split PR?

Yes, wrapping up unit tests right now on this one for the mechanics and then we can do a follow up that will add in the default sets of caps to our bootstrap policies where appropriate.

pweil- commented:
Updated with more tests and cleaned up comments.  Ready for review.

@pmorie @liggitt - could I get some eyes on this one when you get a chance?

smarterclayton commented:
Changes LGTM but want another sign off.

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2597,7 +2597,19 @@ type SecurityContextConstraints struct {
 
 	// AllowPrivilegedContainer determines if a container can request to be run as privileged.
 	AllowPrivilegedContainer bool `json:"allowPrivilegedContainer" description:"allow containers to run as privileged"`
+	// Required capabilities are the capability set that must be set or defaulted into
```
How about `RequiredCapabilities are defaulted into containers.` ?

pweil- replied:
done

pweil- commented:
@pmorie @liggitt @deads2k  - any other comments on the mechanics here?  I've run through my tests and it appears to be defaulting and restricting correctly.  We can iterate on the actual values of the defaults separately (they are in https://github.com/openshift/origin/issues/6627)

pweil- commented:
[test]

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/validation/validation_test.go:
```diff
@@ -4171,6 +4171,18 @@ func TestValidateSecurityContextConstraints(t *testing.T) {
+	allowedCapListedInRequiredDrop.RequiredCapabilities = api.Capabilities{
+		Drop: []api.Capability{"foo"},
+	}
+	allowedCapListedInRequiredDrop.AllowedCapabilities = []api.Capability{"foo"}
```
what about a test case where a capability is in both add, drop, and allowed?

pweil- replied:
the validation doesn't care about that.  It tests required.Add -> required.Drop and Allowed -> required.Drop.  It is valid to have it in allowed and required.Add though I don't see why someone would do that.

pmorie replied:
Okay, I'll buy that for now.  If people find it incredibly confusing we can add a validation later.

liggitt replied:
can we add a validation later if that would invalidate existing sccs?

pweil- replied:
I don't think we need validation for this but if someone really wants it I can add it.  If you put something in both allowed and required.add it is just redundant but it isn't invalid.  It is invalid to have something in (allowed or required.add) AND required.drop which is what we're validating.

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -167,7 +175,12 @@ func (s *simpleProvider) CreateContainerSecurityContext(pod *api.Pod, container
 	}
 
-	// No need to touch capabilities, they will validate or not.
+	caps, err := s.capabilitiesStrategy.Generate(pod, container)
```
Are there going to be any other types of strategies that apply in lieu of the must run as strategy?

pweil- replied:
not that I can think of

pweil- commented:
re[test]

smarterclayton commented:
Any other comments on this?

On Wed, Jan 13, 2016 at 8:40 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8540/
> )
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6470#issuecomment-171497244>.
>

soltysh commented:
/sub

pweil- commented:
hrm, flake on a route conversion re[test]

liggitt commented:
flake is https://github.com/openshift/origin/issues/6656

smarterclayton commented:
This is approved for merge given previously assessed risk

pmorie commented:
This LGTM.

pweil- commented:
@liggitt - If you're satisfied with the comments above (and code) then this seems to be ready.

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/types.go:
```diff
@@ -2143,7 +2143,19 @@ type SecurityContextConstraints struct {
+	// 3.  Capabilities in RequiredCapabilities.Drop will be dropped from the container
+	// 4.  Containers that request to add a capability that is required to be dropped will fail
+	//		validation.
+	RequiredCapabilities Capabilities
```
if we let pods drop capabilities that are in the requiredCapabilities.Add list, then "requiredCapabilities" is an odd name. "defaultCapabilities"?

pweil- replied:
sgtm

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/validation/validation.go:
```diff
@@ -2105,9 +2105,52 @@ func ValidateSecurityContextConstraints(scc *api.SecurityContextConstraints) err
+
+// hasCap checks for needle in haystack.
+func hasCap(needle api.Capability, haystack []api.Capability) bool {
+	needleToLower := strings.ToLower(string(needle))
```
are all capabilities case-insensitive?

pweil- replied:
In the docker values, yes.  It shouldn't matter if we pass upper or lower

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas.go:
```diff
@@ -0,0 +1,169 @@
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// mustRunAs implements the RunAsUserSecurityContextConstraintsStrategy interface
```
old doc, and mustRunAs is a weird name for this

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -228,20 +241,7 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
-		for _, cap := range sc.Capabilities.Add {
-			found := false
-			for _, allowedCap := range s.scc.AllowedCapabilities {
-				if cap == allowedCap {
```
we previously required capabilities match allowedCapabilities with exact case... is there a good reason to loosen that?

pweil- replied:
I loosened it while testing to eliminate two things: ensuring the comparisons of allowed/required.add -> required.drop were flagging values despite case differences and to ensure that you wouldn't end up with duplicates in the same add/drop list.

liggitt replied:
and all consumers of capabilities are happy with the lower-cased versions?

pweil- replied:
I'll just toUpper them.  It works fine for the docker values but the docker documentation has them written in caps too.  These values are inherently container runtime dependent though.  For instance you don't say you want CAP_SYS_ADMIN when adding the cap in docker you just use SYS_ADMIN.  Another example - docker has a special case for the ALL value.

liggitt replied:
My concern is that we're setting API behavior we'll have to support long-term, based on a single runtime's behavior. Also, we try to avoid normalization of user input in the API... I'd rather make them say what they mean, but if no one else is bothered by it (or cannot envision a time when we'll need to preserve the case of what the user specifies for a container runtime), then toUpper is better

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+func capabilityFromStringSlice(slice []string) []api.Capability {
+	if len(slice) == 0 {
```
may not matter, but I would expect `nil -> nil` and `[] -> []`

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas_test.go:
```diff
@@ -0,0 +1,369 @@
+			expectedCaps: &api.Capabilities{
+				Add: []api.Capability{"bar", "foo"},
+			},
+		},
```
test case mismatches:
* container drop of "FOO" prevents an add of a requiredCapabilities.Add=[foo]
* container add of "FOO" is deduped with a requiredCapabilities.Add=[foo]
* container drop of "FOO" is deduped with a requiredCapabilities.Drop=[foo]

test deduping:
* requiredCapabilities.Add=[foo,foo,FOO] is deduped
* requiredCapabilities.Drop=[foo,foo,FOO] is deduped
* containerCapabilities.Add=[foo,foo,FOO] is deduped
* containerCapabilities.Drop=[foo,foo,FOO] is deduped

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas_test.go:
```diff
@@ -0,0 +1,369 @@
+				Drop: []api.Capability{"bar"},
+			},
+		},
+		"defauting adds and drops, dropping a required add": {
```
typo `defauting`

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas_test.go:
```diff
@@ -0,0 +1,369 @@
+			expectedCaps: &api.Capabilities{
+				Add:  []api.Capability{"bar", "baz"},
+				Drop: []api.Capability{"abc", "foo"},
+			},
```
test case-insensitivity

liggitt commented on pkg/security/admission/admission_test.go:
```diff
@@ -27,6 +27,126 @@ func NewTestAdmission(store cache.Store, kclient client.Interface) kadmission.In
+		},
+		// UC 3: if an SCC requires a cap then it should accept the pod request
+		// to add the cap.
+		"should accept cap add when in requied": {
```
required

pweil- commented:
Updates:

1.  Rename to `InitialCapabilities`
1.  Rename of strategy
1.  Added back case sensitivity - this seems the safest route because these are container runtime specific values
1.  Added tests for case sensitivity for add/drop generation and validation
1.  Added tests to ensure deduping of matched case items during generation of adds and drops
1.  Fixed typos

@liggitt @pmorie

pweil- commented:
re[test]

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2597,7 +2597,19 @@ type SecurityContextConstraints struct {
+	// 3.  Capabilities in InitialCapabilities.Drop will be dropped from the container
+	// 4.  Containers that request to add a capability that is required to be dropped will fail
+	//		validation.
+	InitialCapabilities Capabilities `json:"initialCapabilities" description:"capabilities that will be added or dropped"`
```
This name makes it seem like capabilities can change over time. "initial" needs a time context to be understood, and it's a little odd because the sequence is: 1) pod spec submitted to the API may specify capabilities, 2) admission drops some capabilities, adds some capabilities unless dropped by the pod, 3) object is created and capabilities can never be changed.

What about "assignedCapabilities"? They are what admission will assign to the pod at creation time, and the pod author can explicitly opt out of some of the added capabilities by dropping them in their submitted pod spec.

eparis replied:
I know it is bike shedding, but 'assigned' sounds to me like it would blindly overwrite things. Which it only sort of does. You don't end up with the 'assigned' caps. You end up with some amalgamation.

I'm not arguing for Initial either, I'll try to think of something to propose instead of just being a negative nancy...

pweil- replied:
Last IRC suggestion was:

`allowedCapabilities []kapi.Capbility` - what you're allowed to add (already exists)
`defaultAddCapabilitiess []kapi.Capbility`  - what you'll be given by default unless you drop one of them explicitly
`requiredDropCapabilitiess []kapi.Capbility`  - you must drop these

@smarterclayton @eparis @liggitt - satisfactory?

smarterclayton replied:
allowedAddCapabilities

liggitt replied:
Can't change that one, it already exists

smarterclayton replied:
I'm not in love with the names but we don't have time to bike shed.  Names are approved

pweil- commented:
@liggitt @eparis - updated, PTAL

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas.go:
```diff
@@ -0,0 +1,166 @@
+// 		container specifically is dropping the cap) and container requested adds
+// 2.  a capabilities.Drop set containing all the required drops and container requested drops
+func (s *defaultCapabilities) Generate(pod *api.Pod, container *api.Container) (*api.Capabilities, error) {
+	requiredAdd := makeCapSet(s.defaultAddCapabilities)
```
s/requiredAdd/defaultAdd/

pweil- replied:
fixed

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas.go:
```diff
@@ -0,0 +1,166 @@
+	containerAdd := sets.NewString()
+	containerDrop := sets.NewString()
+
+	if scHasCapAdd(container.SecurityContext) {
```
are we using those helper functions anywhere else? this would probably be easier to read as:
```
containerAdd := sets.NewString()
containerDrop := sets.NewString()
if container.SecurityContext != nil && container.SecurityContext.Capabilities != nil {
  containerAdd = makeCapSet(container.SecurityContext.Capabilities.Add)
  containerDrop = makeCapSet(container.SecurityContext.Capabilities.Drop)
}
```

pweil- replied:
nope, will remove all the helpers

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/capabilities/mustrunas.go:
```diff
@@ -0,0 +1,166 @@
+		containerDrop = makeCapSet(container.SecurityContext.Capabilities.Drop)
+	}
+
+	// remove any required adds that the container is specifically dropping
```
default adds

pweil- replied:
fixed

pweil- commented:
re[test]

pweil- commented:
[test]

pweil- commented:
@liggitt anything else on the revised code that looks fishy?  I'm just fighting with jenkins at this point.

liggitt commented:
nope, LGTM

smarterclayton commented:
[merge]

smarterclayton commented:
[merge]
